### ISSUE TITLE
User-submitted performance & quality reports

### ISSUE THREAD

**Issue by EwoutH** [AOMediaCodec/SVT-AV1#98]
In this issue users can leave their own performance reports and observations.

**Comment by EwoutH** [comment]
A quick speed test I ran overnight between a few recent builds from the master branch. The 70% performance increase claimed by #69 was reproduced (75,7%), while the bit-rate dropped by 6,5%.

| Build    | Build date | Time (ms) | Speed (fps) | Bitrate (kb/s) |
|----------|-----------:|----------:|------------:|---------------:|
| 77174da0 |      02-21 |    938296 |       0,128 |         298,91 |
| a39acb2f |      02-21 |    946272 |       0,127 |         298,91 |
| aee586f1 |      02-20 |   1662581 |       0,072 |         319,68 |
| 476fc452 |      02-19 |   1703176 |       0,070 |         319,68 |

Also a quick test on tiling performance, adding 4 tiles increased the bitrate by 16,9%. I will run a more extensive test soon.

| Build    | Build date | Time (ms) | Speed (fps) | Bitrate (kb/s) | Exta options    |
|----------|-----------:|----------:|------------:|---------------:|-----------------|
| 77174da0 |      02-21 |    953014 |       0,126 |         349,54 | -tile-columns 4 |
| 77174da0 |      02-21 |    938296 |       0,128 |         298,91 |                 |

- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** ` -enc-mode 0 -n 120`
- **System:** i5-4590 (4c/4t, AVX2), 16GB DDR3, Windows 10 64-bit

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit?usp=sharing).

**Comment by EwoutH** [comment]
A quick speed test comparing the different encoder modes. `-enc-mode 4` is slower than `-enc-mode 3` (#100), and `-enc-mode 1` crashed (#99).

| Setting     | Time (ms) | Speed (fps) | Bitrate (kb/s) |
|-------------|----------:|------------:|---------------:|
| -enc-mode 4 |     64663 |       1,856 |         346,95 |
| -enc-mode 3 |     45612 |       2,631 |         345,55 |
| -enc-mode 2 |    178345 |       0,673 |         314,02 |
| -enc-mode 1 |     crash |             |                |
| -enc-mode 0 |    947358 |       0,127 |         298,91 |

- **Build:** 425ba777 (2019-02-22)
- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-n 120`
- **System:** i5-4590 (4c/4t, AVX2), 16GB DDR3, Windows 10 64-bit

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=711455734).

**Comment by EwoutH** [comment]
A more extensive test measuring impact of tiling on the bitrate (kb/s). Note that adding columns increases the bitrate more than adding rows, is this expected behavior?

Encoding speed was equal around 0,68 fps.

| Bitrate         | -tile-rows 0 | -tile-rows 1 | -tile-rows 2 |
|-----------------|-------------:|-------------:|-------------:|
| **-tile-columns 0** |       314,02 |       318,48 |       327,22 |
| **-tile-columns 1** |       336,15 |       341,05 |       354,43 |
| **-tile-columns 2** |       362,07 |       372,46 |              |
| **-tile-columns 3** |       399,21 |       421,10 |              |

- **Build:** 425ba777 (2019-02-22)
- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-enc-mode 2 -n 120`
- **System:** i5-4590 (4c/4t, AVX2), 16GB DDR3, Windows 10 64-bit

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=1883908341).

**Comment by chkngit** [comment]
measuring rate only might not give you the full picture, you probably might wanna add a PSNR or SSIM along, or get the BD rate cost.  
the results would depend on the  spatial characteristic of  the  content it self, and to the number of tiles you add on each dimension

**Comment by EwoutH** [comment]
@chkngit #97 would help a lot with that :)

**Comment by 1480c1** [comment]
Could you either upload the clip used in the test or put a link to the source? I'm interested in doing the same tests, however I do not know what Morocco clip you have used and thus am hesitant to do the same tests to compare.

**Comment by EwoutH** [comment]
@1480c1 It's this [clip](https://www.youtube.com/watch?v=hVvEISFw9w0).

I used youtube-dl to download the 4K version, then scaled it with FFmpeg to different resolutions .y4m (1080p used in above tests).

**Comment by 1480c1** [comment]
> @1480c1 It's this [clip](https://www.youtube.com/watch?v=hVvEISFw9w0).
> 
> I used youtube-dl to download the 4K version, then scaled it with FFmpeg to different resolutions .y4m (1080p used in above tests).

Thank you, will test this when I can and upload the results in the same format.

**Comment by gnafuthegreat** [comment]
> Note that adding columns increases the bitrate more than adding rows, is this expected behavior?

That's interesting, because in my experience the opposite is true with aomenc.  I rarely see a big change in size when using tile-columns, but there's usually a notable increase in size using tile-rows.  Of course, that may just come down to a difference in the types of content we've both tested.

**Comment by 1480c1** [comment]
A couple of things I would like to note, turbo-boost. Main reason is because due to the small amount of frames(120) and short encoding times, my cpu turbo boosts for the first few seconds, in which some of the shorter encodes are done, but for some of the longer ones, my cpu goes back down to the base clock. some of the more shorter encoding, it may not be actually representative of real life values, meaning if someone in the future wanting to decide to uses these information for comparing av1 encoders, may be misled by lower bitrates at some of the faster speed.

tldr
Maybe add a disclaimer that these results might not reflect actual speeds when trying to compare to other encoders with actual footage due to the shortened video and maybe include the exact command lines you used to produce your data, it would make reproducing the data set a little bit more easier.

e.g. 
`ffmpeg.exe -i '.\Morocco 8K HDR 60FPS (FUHD)-hVvEISFw9w0.webm' -vf scale=1920:1080 -vframes 120 Morocco.yuv`
`SvtAv1EncApp.exe -n 120 -w 1920 -h 1080 -fps-num 2997 -fps-denom 50 -bit-depth 8 -i Morocco.yuv -enc-mode n`

| Setting |	Time (ms) | Speed (fps) | Bitrate (kb/s)|
| --- | --- | --- | --- |
| -enc-mode 4 | 52018| 2.307| 324.1|
| -enc-mode 3 | 35571| 3.374|	320.850|
| -enc-mode 2 | 143183|	0.838| 300.330|
| -enc-mode 1 | Crash| Crash | Crash |
| -enc-mode 0 | 780136| 0.154| 287.880|
 * **Build:** [afe72f6](https://github.com/OpenVisualCloud/SVT-AV1/commit/afe72f6dbadbe52b239676d3330f9c4ffa692905) (2019-02-22)
 * **Video:** Morocco (1080p 8-bit 4:2:0)
 * **Settings:** ` -n 120 -w 1920 -h 1080 -fps-num 2997 -fps-denom 50 -bit-depth 8 -i Morocco.yuv -enc-mode n`
 * **System:** Intel Core i7-7700k (4c/8t, AVX2), 32GB DDR4, Windows 10 64-bit

[Full details](https://docs.google.com/spreadsheets/d/1KAQPXyh7UW_QiciD9gYq-3Y2boTr1nNniGHSoPvR5qw/edit?usp=sharing).

**Comment by cdgriffith** [comment]
| Setting 	  |  Time (ms) |	Speed (fps) |	Bitrate (kb/s) |
| --- |	--- 	   | ---	        |   --- |
| -enc-mode 4 |	25817 	   | 4.648 	        |   328.84 |
| -enc-mode 3 |	22164 	   | 5.414 	        |   328.45 |
| -enc-mode 2 |	81928 	   | 1.465 	        |   308.14 |
| -enc-mode 1 |	193687 	   | 0.620 	        |   307.07 |
| -enc-mode 0 |	408742 	   | 0.294 	        |   292.63 |

* **Build:** https://github.com/OpenVisualCloud/SVT-AV1/commit/7b6733d57ce910303a7415a60e184211f8aaf17a (2019-02-22)
* **Video:** Morocco (1080p 8-bit 4:2:0)
    * **Settings:** ` -n 120 -w 1920 -h 1080 -fps-num 2997 -fps-denom 50 -bit-depth 8 -i Morocco.yuv -enc-mode n`
* **System:** Intel Core i9-9900k (8c/16t, AVX2), 32GB DDR4, Windows 10 64-bit

CPU was pegged for mode 3 and 4, modes 1 and 2 around 50\~60%.  Mode 0 was 100% for a few seconds for the first frame then dropped dramatically to 30\~40%.

**Comment by EwoutH** [comment]
Great to see more people using and testing SVT-AV1!

Some suggestions for tests (all my machines are busy with dav1d and rav1e at the moment).
 - #113 CPU core scaling: `-lp` 1 to n (where n is the number of cores of your machine)
 - #114 Changed M2 settings: `-enc-mode 2` on master and the PR
 - #115 Adaptive QP Scaling
 - #108 AVX2 performance (ideally over multiple speed levels)
 - #94 AVX2 performance (ideally over multiple speed levels)

**Comment by EwoutH** [comment]
Did a scaling test with `-lp 12` to `-lp 1`. Raw results:
 
| Speed (fps) | enc-mode 3 | enc-mode 2 | enc-mode 1 |
|------------:|-----------:|-----------:|-----------:|
|          12 |      2,476 |      0,737 |      0,300 |
|          11 |      2,426 |      0,738 |      0,306 |
|          10 |      2,119 |      0,658 |      0,269 |
|           9 |      2,074 |      0,664 |      0,273 |
|           8 |      1,719 |      0,566 |      0,232 |
|           7 |      1,690 |      0,558 |      0,230 |
|           6 |      1,303 |      0,441 |      0,183 |
|           5 |      1,270 |      0,431 |            |
|           4 |      0,936 |      0,324 |            |
|           3 |      0,871 |      0,289 |            |
|           2 |      0,474 |      0,163 |            |
|           1 |       fail |       fail |       fail |

`-lp 1` failed on all enc-modes with a memory allocation error, which is strange because less threads should mean less memory usage. 
`Could not allocate enough memory for channel 1`

The benefits from adding a thread are much smaller than those from adding a core, as observable in following table (results are normalized to `-lp 2)`.

| Speed (%) | enc-mode 3 | enc-mode 2 |
|----------:|-----------:|-----------:|
|        12 |      87,1% |      75,4% |
|        11 |      93,1% |      82,3% |
|        10 |      89,4% |      80,7% |
|         9 |      97,2% |      90,5% |
|         8 |      90,7% |      86,8% |
|         7 |     101,9% |      97,8% |
|         6 |      91,6% |      90,2% |
|         5 |     107,2% |     105,8% |
|         4 |      98,7% |      99,4% |
|         3 |     122,5% |     118,2% |
|         2 |     100,0% |     100,0% |

The uneven thread counts add a core with one thread, while the even thread counts only add a second thread to an already utilized core.

If we sort from low to high and only look at the even thread counts (with fully utilized cores), we see pretty impressive core scaling. Note that the core scaling is better on higher enc-modes.

| Speed (%) | enc-mode 3 | enc-mode 2 |
|----------:|-----------:|-----------:|
|         2 |     100,0% |     100,0% |
|         4 |      98,7% |      99,4% |
|         6 |      91,6% |      90,2% |
|         8 |      90,7% |      86,8% |
|        10 |      89,4% |      80,7% |
|        12 |      87,1% |      75,4% |

![](https://svgshare.com/i/BNv.svg)

- **Build:** e74b841 (2019-02-23)
- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-n 120`
- **System:** Ryzen 5 1600 (6c/12t, AVX2), 16GB DDR4, Windows 10 64-bit
- **Command:** `FOR %a in (3,2,1) DO (FOR %b in (0,12,11,10,9,8,7,6,5,4,3,2,1) DO SvtAv1EncApp -i Morocco_1080p_YT_240,y4m -enc-mode %a -lp %b -n 120)`

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=555390849).

**Comment by EwoutH** [comment]
Quick test of #114. Against master, the speed of `-enc-mode 2` increased with 15,8% while the bitrate decreased with 3,2%. Unfortunately, I don't have Psnr (or other efficiency metrics) integrated in my test process yet (#97 would help).

|                | df11eb0 | 0f77af8 |
|----------------|--------:|--------:|
| Speed (fps)    |   0,587 |   0,680 |
| Bitrate (kb/s) |  328,63 |  317,98 |

|                | df11eb0 | 0f77af8 |
|----------------|--------:|--------:|
| Speed (fps)    |  100,0% |  115,8% |
| Bitrate (kb/s) |  100,0% |   96,8% |

![](https://svgshare.com/i/BRR.svg)

- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-enc-mode 2 -n 240`
- **System:** i5-4590 (4c/4t, AVX2), 16GB DDR3, Windows 10 64-bit

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=2045023166).

**Comment by EwoutH** [comment]
Did a quick test of thread scaling. Only 200 frames, but a longer test with 1000 frames seems to follow a similar patern. Aggressive scaling up to 14 threads, then stagnation until 28 threads, and then aggressive scaling until 42 threads. It's almost if at 30 threads a new pool of memory/cache is used (the second half?).

![](https://svgshare.com/i/BSC.svg)

- **Build:** ad88bc6 (2019-02-25)
- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-enc-mode 3 -n 200`
- **System:** Intel (28c/56t, AVX2), 128GB DDR4, Linux Mint

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=1797152795).

**Comment by EwoutH** [comment]
Almost the exact same curve on the first 16,67 seconds of the clip (`-n 1000`):
![](https://svgshare.com/i/BU5.svg)

**Comment by tianjunwork** [comment]
@EwoutH -lp config acts this way:
The granularity is by thread(logical processor) indicated by /proc/cpuinfo processor id.
So -lp 0 use processor 0, -lp 1 adds processor 1, and so on.
You can use "lstopo" to see how the processors are distributed on 28 cores.
So with the increasing -lp, the program first use one thread per core on socket#0 ==> then use the other thread on socket#0 ==> use one thread per core on socket#1 ==> use the other thread on socket#1.

Yes, your understanding is correct. "up to 14 threads new cores are used each time, and then 16-28 only adds threads. Then 30-42 is ran on new cores again, and 44-50 only adds more threads."

**Comment by EwoutH** [comment]
@tianjunwork Thanks for confirming the behavior. I first though I was using a single-socket 28c machine, but it's actually 2x 14c (it's not my PC, only got remote access). 

It's strange that on the Ryzen (Windows?) CPU every second thread uses an already used CPU-core, while on the Intel (Linux?) CPU every new thread uses a new core (up to 14). Or is this an OS difference in how the cores are numbered?

----
@hassount I measures a bitrate increase of 19% to 25% between 23-02 and today with the same encoder settings. Is this expected? ([files here](https://drive.google.com/drive/folders/1mfroRgigxdPpJJiS9BX15FMV-g2Xcmfg), 08 is df11eb0 and 09 is 04f8e2d)

Bitrate (kb/s) | df11eb0 | 04f8e2d |  
-- | -- | -- | --
-enc-mode 4 | 339,45 | 421,10 | 124,05%
-enc-mode 3 | 334,58 | 418,30 | 125,02%
-enc-mode 2 | 328,63 | 400,21 | 121,78%
-enc-mode 1 | 331,05 | 395,49 | 119,47%

**Comment by tianjunwork** [comment]
@EwoutH Thread setting API and internal scheduling implementation is different on OS.
NOTE on windows:
https://docs.microsoft.com/en-us/windows/desktop/procthread/processor-groups
By default, an application is constrained to a single group. An application that requires the use of multiple groups so that it can run on more than 64 processors must explicitly determine where to run its threads and is responsible for setting the threads' processor affinities to the desired groups.
Since on windows thread affinity is set by **group** on multi-group system, on 40(thread)x2 system, -lp 50(larger than 40) will config the encoder to run on actually all threads(80).

**Comment by EwoutH** [comment]
My findings about 47af208 (#185) compared to my previous tests:
- enc-mode 8 is slower than 7
- enc-mode 7 to 4 are slower than on 2019-03-04
- enc-mode 7 to 0 are faser than on 2019-03-04

| Speed (fps) |  0304 | 0319  |
|-------------|------:|-------|
| -enc-mode 8 |       | 15,91 |
| -enc-mode 7 | 22,55 | 16,86 |
| -enc-mode 6 | 16,25 | 14,56 |
| -enc-mode 5 | 13,26 | 12,24 |
| -enc-mode 4 | 8,56  | 6,24  |
| -enc-mode 3 | 2,97  | 3,47  |
| -enc-mode 2 | 0,87  | 2,13  |
| -enc-mode 1 | 0,51  | 1,28  |
| -enc-mode 0 | 0,32  | 0,63  |

**Comment by hassount** [comment]
@EwoutH 
Thank you for the testing results. something seems way off with the speed numbers, maybe something happened during the merge let me take a look.

**Comment by hassount** [comment]
@EwoutH 
I am a little confused, the same content generates such a large difference on a the google platform vs the dual Xeon. The number of frames is too small for these speed tests as most of the time is spent reading the yuv and or ramping up the cpu especially with the 96 cores instance. We generally use ~1000 frames at least. I will share what we got on our side very soon.

**Comment by EwoutH** [comment]
@hassount I ran with 1000 frames, enc-mode 7 to 0 are now faster on Google, except mode 4, and enc-mode 8 is still slower than 7 (see links above).

**Comment by hassount** [comment]
@EwoutH 
Thank you for re running, yes I noticed the speed drop. we'll do some testing for 1080p on a 96 core server to see if we can reproduce such on our end. Do you know where I could get access to the video content ? I might have missed a link you posted before

**Comment by EwoutH** [comment]
@hassount https://media.xiph.org/, I used the Sintel trailer (1080p y4m)

**Comment by EwoutH** [comment]
I just ran some perf reports on my system. Here the profiles for everyone who might find it interesting!

- [perf-report-svtav1-0521-m8.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3204526/perf-report-svtav1-0521-m8.txt)
- [perf-report-svtav1-0521-m6.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3204529/perf-report-svtav1-0521-m6.txt)
- [perf-report-svtav1-0521-m4.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3204528/perf-report-svtav1-0521-m4.txt)
- [perf-report-svtav1-0521-m2.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3204527/perf-report-svtav1-0521-m2.txt)

_Config_

- **Build:** e1a6158 (2019-05-20)
- **Video:** Morocco (1080p 8-bit 4:2:0)
- **System:** Intel (28c/56t, AVX2), 128GB DDR4, Linux Mint
- **Commands:** [SVT-AV1 commands 0520 linux SSH server perf record.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3204540/SVT-AV1.commands.0520.linux.SSH.server.perf.record.txt)

**Comment by EwoutH** [comment]
Just a [general benchmark](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=1378779628) of encoding speed in the last month.

| Build       |    bef6750 |    c097dfb |        |
|-------------|-----------:|-----------:|-------:|
| Build date  | 2019-04-27 | 2019-05-23 |        |
| -enc-mode 8 |     53,860 |     57,554 | 106,9% |
| -enc-mode 7 |     25,097 |     27,021 | 107,7% |
| -enc-mode 6 |     18,464 |     19,386 | 105,0% |
| -enc-mode 5 |      9,536 |     10,447 | 109,6% |
| -enc-mode 4 |      4,387 |      4,619 | 105,3% |
| -enc-mode 3 |      3,260 |      3,369 | 103,3% |
| -enc-mode 2 |      1,270 |      1,294 | 101,9% |
| -enc-mode 1 |      0,587 |      0,598 | 101,9% |
| -enc-mode 0 |      0,194 |      0,198 | 102,1% |

**Comment by EwoutH** [comment]
New perf reports (same system as above, 1d85a30):

- [perf-report-svtav1-0606-m0.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3263532/perf-report-svtav1-0606-m0.txt)
- [perf-report-svtav1-0606-m2.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3263533/perf-report-svtav1-0606-m2.txt)
- [perf-report-svtav1-0606-m4.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3263534/perf-report-svtav1-0606-m4.txt)
- [perf-report-svtav1-0606-m6.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3263535/perf-report-svtav1-0606-m6.txt)
- [perf-report-svtav1-0606-m8.txt](https://github.com/OpenVisualCloud/SVT-AV1/files/3263536/perf-report-svtav1-0606-m8.txt)

**Comment by EwoutH** [comment]
[Did a test](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=899117475) with different versions of the GCC compiler today. GCC 8.3 was about 2% faster than GCC 7.3, between 8.3 and 9.1 no significant performance difference was observed.

| Build       | GCC 7.3 | GCC 8.3 | GCC 9.1 | 7.3 vs 8.3 | 8.3 vs 9.1 |
|-------------|--------:|--------:|--------:|-----------:|-----------:|
| -enc-mode 8 |  48,385 |  49,291 |  49,715 |     101,9% |     100,9% |
| -enc-mode 7 |  27,361 |  27,834 |  27,833 |     101,7% |     100,0% |
| -enc-mode 6 |  17,448 |  17,791 |  17,596 |     102,0% |      98,9% |
| -enc-mode 5 |  10,863 |  11,158 |  11,074 |     102,7% |      99,2% |
| -enc-mode 4 |   4,501 |   4,604 |   4,638 |     102,3% |     100,7% |
| -enc-mode 3 |   3,261 |   3,306 |   3,314 |     101,4% |     100,2% |
| -enc-mode 2 |   0,771 |   0,786 |   0,788 |     101,9% |     100,3% |
| -enc-mode 1 |   0,411 |   0,419 |   0,415 |     101,9% |      99,0% |

**Comment by 1480c1** [comment]
Here's one done by someone else about the 0.8.1 release https://www.reddit.com/r/AV1/comments/f0m4tb/svtav1_081_vmaf_test/

![graph](https://preview.redd.it/fkubtpgngmf41.png?width=960&crop=smart&auto=webp&s=fec2c75615c301563eb5c3053b9e6fdf6f377542)

@hassount, there seems to be almost no difference between enc mode 0 and 1, 3 and 4, and barely any between 6 and 7 except for time

**Comment by 1480c1** [comment]
http://compression.ru/video/codec_comparison/hevc_2019/#hq_report

Seems that SVT-AV1 was only beat by aom for average bitrate at same quality
